pkg/server/cluster_server.go

```
@@ -45,16 +62,70 @@ type clusterServer struct {
+		return nil, errors.Wrapf(err, "Failed to create Kubernetes rest client")
 	}
 
-	mc := v1.NewForConfigOrDie(restConfig)
+	mc, err := v1.NewForConfig(restConfig)
+	if err != nil {
+		return nil, errors.Wrapf(err, "creating mc client")
+	}
+	oc, err := oconfigv1.NewForConfig(restConfig)
+	if err != nil {
+		return nil, errors.Wrapf(err, "creating config client")
+	}
 	return &clusterServer{
 		machineClient:  mc,
+		configClient:   *oc,
 		kubeconfigFunc: func() ([]byte, []byte, error) { return kubeconfigFromSecret(bootstrapTokenDir, apiserverURL) },
 	}, nil
 }
 
+// remoteAddrIsProvisioned returns true if the given IP address responds on a known port.
+// If that's true, the machine is already up and should not be served Ignition.
+func remoteAddrIsProvisioned(remoteAddr string) (string, error) {
+	var wg sync.WaitGroup
+	var provisionedPort string
+	wg.Add(len(machineProvisionedPorts))
+	for _, port := range machineProvisionedPorts {
+		go func(port string) {
+			conn, err := net.DialTimeout("tcp", net.JoinHostPort(remoteAddr, port), time.Second * machineProvisionedTimeoutSecs)
+			defer conn.Close()
+			if err != nil {
+				return
+			}
+			provisionedPort = port
+		}(port)
+	}
+	wg.Wait()
+	return provisionedPort, nil
+}
+
+// remoteAddrIsFromPod returns true if the given remote IP
+// is from a pod.  We don't want to allow
+// pods running on a node to potentially access Ignition; it should
+// only be read by the Ignition software itself when the machine
+// boots up, before it joins the cluster.
+func remoteAddrIsFromPod(configClient oconfigv1.ConfigV1Client, remoteAddr string) (string, error) {
+	network, err := configClient.Networks().Get("cluster", metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+	remoteIP := net.ParseIP(remoteAddr)
+	if remoteIP == nil {
+		return "", nil
+	}
+	for _, n := range network.Status.ClusterNetwork {
+		_, ipr, err := net.ParseCIDR(n.CIDR)
+		if err != nil {
+			continue
```

nit: debug log the error rather than swallow